src/lazy_index_map.rs

```diff
@@ -58,14 +74,24 @@ where
         K: Borrow<Q> + PartialEq<Q>,
         Q: Hash + Eq,
     {
-        let map = self.map.get_or_init(|| {
-            self.vec
-                .iter()
-                .enumerate()
-                .map(|(index, (key, _))| (key.clone(), index))
-                .collect()
-        });
-        map.get(key).map(|&i| &self.vec[i].1)
+        let vec_len = self.vec.len();
+        // if the vec is longer than the threshold, we use the hashmap for lookups
+        if vec_len > HASHMAP_THRESHOLD {
+            self.get_map().get(key).map(|&i| &self.vec[i].1)
+        } else {
+            // otherwise we find the value in the vec
+            // we assume the most likely position for the match is at `last_find + 1`
+            let first_try = self.last_find.load(Ordering::Relaxed) + 1;
+            for i in first_try..first_try + vec_len {
+                let index = i % vec_len;
+                let (k, v) = &self.vec[index];
+                if k == key {
+                    self.last_find.store(index, Ordering::Relaxed);
```

Comment:
Staring at this, I sort of wonder if this `last_find` logic should be moved into a function `pub fn get_hint<Q: ?Sized>(&self, key: &Q, last_find: Option<usize>) -> Option<(&V, usize)>` which returns the hint for the next lookup.

That way the hint can be a local stack variable and not an atomic part of the datastructure. But it does push the complexity downstream.

Going further, maybe this is better abstracted into a a `struct OrderedGetter { map: &LazyIndexMap, last_find: usize }` which looks in the map with a `pub fn get<Q: ?Sized>(&mut self, key: &Q) -> Option<&V>` - i.e. `&mut` - so by doing this is explicitly optimized for the order-preserving read.

Reply:
We spoke about this offline; will keep this as-is for now but maybe can remove the atomic later.

Reply:
We decided on a call to stick with this to avoid churn in pydantic-core.